src/engines/ws.ts

```diff
@@ -0,0 +1,205 @@
+					decoded != null
+				) {
+					this.handleRpcResponse(decoded);
+				} else {
+					this.setStatus(
+						ConnectionStatus.Error,
+						new UnexpectedServerResponse(decoded),
+					);
+				}
+			});
+		});
+
+		this.ready = ready;
+		return await ready.then(() => {
+			this.socket = socket;
+			this.pinger = new Pinger(30000);
+			this.pinger?.start(() => this.rpc({ method: "ping" }));
+		});
+	}
+
+	async disconnect(): Promise<void> {
+		this.connection = {};
+		await this.ready?.catch(() => {});
+		this.socket?.close();
+		this.ready = undefined;
+		this.socket = undefined;
+
+		await Promise.any([
+			this.requireStatus(ConnectionStatus.Disconnected),
+			this.requireStatus(ConnectionStatus.Error),
+		]);
+	}
+
+	async rpc<
+		Method extends string,
+		Params extends unknown[] | undefined,
+		Result,
+	>(request: RpcRequest<Method, Params>): Promise<RpcResponse<Result>> {
+		await this.ready;
+		if (!this.socket) throw new ConnectionUnavailable();
+
+		// It's not realistic for the message to ever arrive before the listener is registered on the emitter
+		// And we don't want to collect the response messages in the emitter
+		// So to be sure we simply subscribe before we send the message :)
+
+		const id = getIncrementalID();
+		const response = this.emitter.subscribeOnce(`rpc-${id}`);
+		this.socket.send(this.encodeCbor({ id, ...request }));
+		return response.then(([res]) => {
+			if (res instanceof EngineDisconnected) throw res;
+
+			if ("result" in res) {
+				switch (request.method) {
+					case "use": {
+						this.connection.namespace = request.params?.[0] as string;
+						this.connection.database = request.params?.[1] as string;
+						break;
+					}
+
+					case "signin":
+					case "signup": {
+						this.connection.token = res.result as string;
+						break;
+					}
+
+					case "authenticate": {
+						this.connection.token = request.params?.[0] as string;
+						break;
+					}
+
+					case "invalidate": {
+						this.connection.token = undefined;
+						break;
+					}
+				}
+			}
+
+			return res as RpcResponse<Result>;
+		});
+	}
+
+	// biome-ignore lint/suspicious/noExplicitAny: Cannot assume type
+	handleRpcResponse({ id, ...res }: any): void {
+		if (id) {
+			this.emitter.emit(`rpc-${id}`, [res]);
+		} else if (res.error) {
+			this.setStatus(ConnectionStatus.Error, new ResponseError(res.error));
+		} else {
+			if (isLiveResult(res.result)) {
+				const { id, action, result } = res.result;
+				this.emitter.emit(`live-${id}`, [action, result], true);
+			} else {
+				this.setStatus(
+					ConnectionStatus.Error,
+					new UnexpectedServerResponse({ id, ...res }),
+				);
+			}
+		}
+	}
+
+	get connected(): boolean {
+		return !!this.socket;
+	}
+}
+
+export class Pinger {
+	private pinger?: ReturnType<typeof setTimeout>;
+	private interval: number;
+
+	constructor(interval = 30000) {
+		this.interval = interval;
+	}
+
+	start(callback: () => void): void {
+		this.pinger = setInterval(callback, this.interval);
```

Comment:
I would unref in environments that support it to prevent from keeping the process alive

Reply:
As in `delete this.pinger` beforehand?